### PR TITLE
Support Exporting Mask Rcnn to ONNX

### PR DESCRIPTION
**Pull request by lara-hdr**
In addition to #1401 , these are the last changes to export Maskrcnn to ONNX.
https://github.com/pytorch/pytorch/pull/27566 and https://github.com/pytorch/pytorch/pull/26549.
Like for Faster rcnn, for now we will only be able to export with a batch of 1 (until we support dynamic cases for split).

The test is disabled until the referenced PRs are merged and Resize is implemented in opset 11.

### CODE REVIEW

File: test/test_onnx.py [hunk 1]
```diff
@@ -129,6 +130,24 @@ def forward(self, input, boxes):
...
+
+    @unittest.skip("Disable test until Resize opset 11 is implemented in ONNX Runtime")
+    def test_mask_rcnn(self):
+        images, test_images = self.get_test_images(self)
+
+        model = models.detection.mask_rcnn.maskrcnn_resnet50_fpn(pretrained=True)
+        model.eval()
+        model(images)
+        self.run_model(model, [(images,), (test_images,)])
+
```
Comment (lara-hdr) [inline test/test_onnx.py]: will add a test soon for the roi_heads independently

File: torchvision/models/detection/roi_heads.py [hunk 2]
```diff
@@ -73,7 +74,11 @@ def maskrcnn_inference(x, labels):
     index = torch.arange(num_masks, device=labels.device)
     mask_prob = mask_prob[index, labels][:, None]
 
-    mask_prob = mask_prob.split(boxes_per_image, dim=0)
+    if torchvision._is_tracing():
```
Comment (fmassa) [inline torchvision/models/detection/roi_heads.py]: Can you make the condition be instead `if len(boxes_per_image) == 1`, as you did in #1401 ?

File: torchvision/models/detection/roi_heads.py [hunk 3]
```diff
@@ -303,11 +330,70 @@ def paste_mask_in_image(mask, box, im_h, im_w):
     return im_mask
 
 
+def _onnx_paste_mask_in_image(mask, box, im_h, im_w):
```
Comment (fmassa) [inline torchvision/models/detection/roi_heads.py]: Can you write a test comparing that this function gives the same behavior as the original one?

File: torchvision/models/detection/roi_heads.py [hunk 4]
```diff
@@ -303,11 +330,70 @@ def paste_mask_in_image(mask, box, im_h, im_w):
...
+    im_w = img_shape[1]
+    res_append = torch.zeros(0, im_h, im_w)
+    for i in range(masks.size(0)):
+        mask_res = _onnx_paste_mask_in_image(masks[i][0], boxes[i], im_h, im_w)
+        mask_res = mask_res.unsqueeze(0)
+        res_append = torch.cat((res_append, mask_res))
+    return res_append
+
+
 def paste_masks_in_image(masks, boxes, img_shape, padding=1):
```
Comment (fmassa) [inline torchvision/models/detection/roi_heads.py]: Could you write a test showing that this function behaves the same both during tracing and without?

File: torchvision/models/detection/roi_heads.py [hunk 5]
```diff
@@ -303,11 +330,70 @@ def paste_mask_in_image(mask, box, im_h, im_w):
...
+
+@torch.jit.script
+def _onnx_paste_masks_in_image_loop(masks, boxes, img_shape):
+    im_h = img_shape[0]
+    im_w = img_shape[1]
+    res_append = torch.zeros(0, im_h, im_w)
+    for i in range(masks.size(0)):
+        mask_res = _onnx_paste_mask_in_image(masks[i][0], boxes[i], im_h, im_w)
+        mask_res = mask_res.unsqueeze(0)
+        res_append = torch.cat((res_append, mask_res))
```
[the remark this answers is not in the document]
Comment (lara-hdr) [reply torchvision/models/detection/roi_heads.py]: This is currently a limitation.
Constructing list and appending to it in a loop is not supported; currently in this scenario the list gets eliminated in a post pass in a EliminateDeadCode() call.

We identified the problem and will be working on fixing it.

File: torchvision/models/detection/roi_heads.py [hunk 6]
```diff
@@ -250,10 +255,29 @@ def keypointrcnn_inference(x, boxes):
     return kp_probs, kp_scores
 
 
+def _onnx_expand_boxes(boxes, scale):
+    w_half = (boxes[:, 2] - boxes[:, 0]) * .5
```
[the remark this answers is not in the document]
Comment (lara-hdr) [reply torchvision/models/detection/roi_heads.py]: this was done to set the tensor to a specific type for operations to have operands of the same type. We had a PR in PyTorch that went in to deal with similar cases. I am planning on going through the old PRs and submit a PR that reverts/simplifies the changes that we now support.

File: test/test_onnx.py [hunk 7]
```diff
@@ -294,6 +321,33 @@ def test_faster_rcnn(self):
...
+        boxes *= 50
+        o_im_s = (100, 100)
+        from torchvision.models.detection.roi_heads import paste_masks_in_image
+        from torchvision.models.detection.transform import resize_boxes
+        out = paste_masks_in_image(masks, boxes, o_im_s)
+        jit_trace = torch.jit.trace(paste_masks_in_image,
+                             (masks, boxes,
+                             [torch.tensor(o_im_s[0]),torch.tensor(o_im_s[1])]))
+        out_onnx = jit_trace(masks, boxes, [torch.tensor(100),torch.tensor(100)])
+        assert torch.all(out.eq(out_onnx))
```
Comment (fmassa) [inline test/test_onnx.py]: Can we also add one more test, which runs the `jit_trace` on a different-sized input / boxes, to verify that the control-flow is handled properly?
Something like
```python
assert torch.all(out.eq(out_onnx))

masks = torch.rand(20, 1, 26, 26)
boxes = torch.rand(20, 4)
boxes[:, 2:] += torch.rand(10, 2)
boxes *= 100
o_im_s = (200, 200)

out = paste_masks_in_image(masks, boxes, o_im_s)
out_onnx = jit_trace(masks, boxes, [torch.tensor(200),torch.tensor(200)])
assert torch.all(out.eq(out_onnx))
```

File: test/test_onnx.py [hunk 8]
```diff
@@ -177,6 +178,31 @@ def _init_test_roi_heads_faster_rcnn(self):
             box_score_thresh, box_nms_thresh, box_detections_per_img)
         return roi_heads
 
+    def _init_test_roi_heads_mask_rcnn(self):
```
Comment (fmassa) [inline test/test_onnx.py]: Do we need this for now, or do we have plans to use it?

File: test/test_onnx.py (hunk 7 above)
Comment (lara-hdr) [reply test/test_onnx.py]: I added a separate test but just understood you meant to add it on the same jit_trace which makes more sense. I'll update that

Comment (fmassa) [reply test/test_onnx.py]: Yeah, I meant to reuse the same `jit_trace`, but for different inputs

Comment (lara-hdr) [reply test/test_onnx.py]: added.
(I actually also realized that o_im_s was traced as a constant after adding the test, and fixed it. thanks!)

**Comment by niaoyu** [comment]
> LGTM, thanks a lot Lara!

Hi, i just wonder which version of torch and torchvison would be suitable for export maskrcnn to onnx.

I tried the nightly version of both , but all failed

**Comment by lara-hdr** [comment]
@niaoyu, as explained in https://github.com/pytorch/pytorch/issues/27969 there are some PRs that are not yet merged in PyTorch to export this model.

File: torchvision/models/detection/roi_heads.py [hunk 9]
```diff
@@ -303,11 +330,69 @@ def paste_mask_in_image(mask, box, im_h, im_w):
...
+    boxes = expand_boxes(boxes, scale).to(dtype=torch.int64)
     # im_h, im_w = img_shape.tolist()
     im_h, im_w = img_shape
+
+    if torchvision._is_tracing():
+        return _onnx_paste_masks_in_image_loop(masks, boxes,
+                                               torch.scalar_tensor(im_h, dtype=torch.int64),
+                                               torch.scalar_tensor(im_w, dtype=torch.int64))[:, None]
+
+    boxes = boxes.tolist()
```
Comment (eellison) [inline torchvision/models/detection/roi_heads.py]: is this intended to be here ? the case where torchvision is tracing is already covered above
